File: Samples/TestDataBuilder/TestUserBuilder.cs
namespace Samples.TestDataBuilder
{
	public class TestUserBuilder
	{
		public const string DEFAULT_NAME = "John Smith";
		public const string DEFAULT_ROLE = "ROLE_USER";
		public const string DEFAULT_PASSWORD = "42";
		private string name = DEFAULT_NAME;
		private string password = DEFAULT_PASSWORD;
		private string role = DEFAULT_ROLE;
		private string login;

		private TestUserBuilder()
		{
		}

		public static TestUserBuilder AUser()
		{
			return new TestUserBuilder();
		}

		public TestUserBuilder WithName(string newName)
		{
			name = newName;
			return this;
		}

		public TestUserBuilder WithLogin(string newLogin)
		{
			login = newLogin;
			return this;
		}

		public TestUserBuilder WithPassword(string newPassword)
		{
			password = newPassword;
			return this;
		}

		public TestUserBuilder WithNoPassword()
		{
			password = null;
			return this;
		}

		public TestUserBuilder InUserRole() => InRole("ROLE_USER");

		public TestUserBuilder InAdminRole() => InRole("ROLE_ADMIN");

		public TestUserBuilder InRole(string newRole)
		{
			this.role = newRole;
			return this;
		}

		public TestUserBuilder But() => 
			AUser()
				.InRole(role)
				.WithName(name)
				.WithPassword(password)
				.WithLogin(login);

		public User Build() => new User(name, login, password, role);

		public static User ARegularUser() => AUser().Build();
		
		public static User AnAdmin() => AUser()
			.WithName("Neo")
			.WithLogin("neo")
			.InAdminRole()
			.Build();
	}
}

File: java/src/test/java/ru/kontur/courses/samples/builder/TestUserBuilder.java
package ru.kontur.courses.samples.builder;

public class TestUserBuilder {
    public final String DEFAULT_NAME = "";
    public final String DEFAULT_ROLE = "";
    public final String DEFAULT_PASSWORD = "";
    private String name = DEFAULT_NAME;
    private String password = DEFAULT_PASSWORD;
    private String role = DEFAULT_ROLE;
    private String login;

    private TestUserBuilder() {
    }

    public static TestUserBuilder aUser() {
        return new TestUserBuilder();
    }

    public TestUserBuilder withName(String newName) {
        name = newName;
        return this;
    }

    public TestUserBuilder withLogin(String newLogin) {
        login = newLogin;
        return this;
    }

    public TestUserBuilder withPassword(String newPassword) {
        password = newPassword;
        return this;
    }

    public TestUserBuilder withNoPassword() {
        password = null;
        return this;
    }

    public TestUserBuilder inUserRole() {
        return inRole("ROLE_USER");
    }

    public TestUserBuilder inAdminRole() {
        return inRole("ROLE_ADMIN");
    }

    public TestUserBuilder inRole(String newRole) {
        this.role = newRole;
        return this;
    }

    public TestUserBuilder but() {
        return aUser().inRole(role).withName(name).withPassword(password).withLogin(login);
    }

    public User build() {
        return new User(name, login, password, role);
    }

    public User aRegularUser() {
        return aUser().build();
    }

    public User anAdmin() {
        return aUser()
                .withName("Neo")
                .withLogin("neo")
                .inAdminRole().build();
    }
}
